### DISCUSSION TITLE
Validate docs agains custom values before adding to (or updating) collection

### DISCUSSION THREAD

**Discussion by islamaliev** [sourcenetwork/defradb#2436] [Feature request]
### The feature you would like

I would be nice to have a way of imposing some constrains on certain doc values for all docs in a collection.
For example, if we want to restrict user's age to be min 18, or we want to allow only a set of values like languages ["EN", "DE", "FR"] (line enum).
```graphql
type User {
    name: String!
    age: Int! @constrain(ge: 18, description: "only for 18 and older")
    language: String @constrain(enum: ["EN", "DE", "FR"], description: "language must be "EN", "DE" or "FR")
}
```
It should work similar to [MongoDB's schema validation](https://www.mongodb.com/docs/manual/core/schema-validation/specify-json-schema/). 

Another nice-to-have would be to add simple arithmetic and dynamic value. This would allow us to set, for example, restriction on a date (like meeting time) to be after current time.
```graphql
type Meeting {
    subject: String!
    time: Date! @constrain(gt: {$now}, description: "meeting time not be in the past")
}
```
Another nice-to-have to allow cross-referencing constrains that check one value agains another one.
```graphql
type Vacation {
    location: String
    startDate: Date! 
    endDate: Date! @constrain(gt: {$doc.startDate}, description: "endDate must be after startDate")
}
```

### Additional context

Another thing to keep in mind is how this will work with p2p and to resolve conflicts. For example, if a doc field is set to be not greater then 10 and 2 users in the network increment the current value 0 by say 7. Locally everything will be valid, but when synching it will be set to 14.

**Comment by fredcarle** [comment]
Since an issue was created for this, I'll close the discussion to keep the reference only in one place.
